feat: add OAuth2 applications

code-asher opened the pull request:
This is **part 1** of https://github.com/coder/coder/issues/11084.

It adds the oauth application configuration part.  The actual token exchange is not implemented yet, probably will use `golang/oauth2`.  As a consequence I have not linked the new page in the sidebar, but it can be accessed by going to `/deployment/oauth2-apps`.  The endpoints are restricted to dev.

This is big enough that I felt like it made sense to get a review on it now even though the token exchange is missing.  It is just a bunch of CRUD, so mostly tedious work like queries and a frontend for those queries.

Two areas in particular I am not sure about:

1. ~Did I do the RBAC part right?~
2. I am not sure I am using features/entitlements correctly.  ~I am not actually preventing adding/editing applications but maybe I should?  I was going to prevent the token exchange though.~ I am doing this now.

Relying on backend validation right now but maybe should do it on the frontend since the backend errors are a bit hard to read.  Also could add that emoji picker to the icon field.

code-asher committed:
Add database tables for OAuth2 applications

These are applications that will be able to use OAuth2 to get an API key
from Coder.

code-asher committed:
Add endpoints for managing OAuth2 applications

These let you add, update, and remove OAuth2 applications.

code-asher committed:
Add frontend for managing OAuth2 applications

code-asher commented:
Not sure about the race tests...will look into it later.

Emyrk commented:
Github allows any user to make an Oauth2 app. Do we expect this to be the case for Coder?

Do we want Coder to not allow personal Oauth2 apps, and only allow global oauth apps?

Should apps be scoped to an org?

Emyrk commented on coderd/database/dbauthz/dbauthz.go:
```diff
@@ -2140,6 +2182,20 @@ func (q *querier) InsertMissingGroups(ctx context.Context, arg database.InsertMi
+	return q.db.InsertOAuth2App(ctx, arg)
+}
+
+func (q *querier) InsertOAuth2AppSecret(ctx context.Context, arg database.InsertOAuth2AppSecretParams) (database.OAuth2AppSecret, error) {
+	if err := q.authorizeContext(ctx, rbac.ActionCreate, rbac.ResourceOAuth2AppSecret); err != nil {
+		return database.OAuth2AppSecret{}, err
+	}
+	return q.db.InsertOAuth2AppSecret(ctx, arg)
+}
```
We should incorporate this in the encrypted db layer.

Parkreiner commented on site/src/pages/DeploySettingsPage/OAuth2AppsSettingsPage/EditOAuth2AppPageView.tsx:
```diff
@@ -0,0 +1,308 @@
+          </TableHead>
+          <TableBody>
+            {isLoadingSecrets && <TableLoader />}
+            {!isLoadingSecrets && (!secrets || secrets?.length === 0) && (
```
Could the second case be simplified down to `!secrets?.length`? I think that would catch both the undefined and 0 cases

code-asher replied:
It definitely could, for some reason I like to always compare `length` to  a number rather than make use of falsey checks.

Parkreiner replied:
Honestly, that's generally what I default to, too – I guess my main concern here was making sure the condition isn't too long, since there's not an easy way to name it without moving the value all the way above the return statement

Parkreiner commented on site/src/pages/DeploySettingsPage/OAuth2AppsSettingsPage/EditOAuth2AppPageView.tsx:
```diff
@@ -0,0 +1,308 @@
+            {!isLoadingSecrets &&
+              secrets &&
+              secrets?.length > 0 &&
+              secrets?.map((secret) => (
```
The second and third boolean checks shouldn't be necessary, because if secrets is undefined, the `?.` will short-circuit the expression and render nothing out

code-asher replied:
Oh right!  Good catch.

Parkreiner commented on site/src/pages/DeploySettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx:
```diff
@@ -0,0 +1,87 @@
+        event.preventDefault();
+        const formData = new FormData(event.target as HTMLFormElement);
+        onSubmit({
+          name: formData.get("name") as string,
```
I get that the type assertions are necessary to exclude the `File` type, but do you think we should still have `null` checks?

code-asher replied:
Honestly I should probably use Formik or whatever, and add the validations too.  I just...dislike Formik haha

I figure `null` will be caught by validation, either on the frontend (if I add it) or the backend.  But from a code complete perspective it does seem like adding `null` here is the right move.

code-asher replied:
Since the page is blocked from being used outside of development for now I am going to punt frontend validation for another PR, if that seems chill.

Parkreiner commented on site/src/pages/DeploySettingsPage/OAuth2AppsSettingsPage/OAuth2AppsSettingsPageView.stories.tsx:
```diff
@@ -0,0 +1,35 @@
+  component: OAuth2AppsSettingsPageView,
+};
+
+export const Loading = {
```
Flagging these, too

buenos-nachos reviewed (commented):
TypeScript looks good! There's a lot of files, though, so I'll do a second review pass later today to check the stories and pull the repo down to kick the tires a bit

code-asher commented on docs/api/enterprise.md:
```diff
@@ -427,6 +427,352 @@ curl -X DELETE http://coder-server:8080/api/v2/licenses/{id} \
+| ------ | ------------------------------------------------------- | ----------- | ----------------------------------------------------------- |
+| 200    | [OK](https://tools.ietf.org/html/rfc7231#section-6.3.1) | OK          | array of [codersdk.OAuth2App](schemas.md#codersdkoauth2app) |
+
+<h3 id="get-oauth2-applications.-responseschema">Response Schema</h3>
```
I think you are right, but I am not sure where it is linked from, if anywhere.  :thinking:

code-asher commented on docs/api/enterprise.md:
```diff
@@ -427,6 +427,352 @@ curl -X DELETE http://coder-server:8080/api/v2/licenses/{id} \
+| ------ | --------------------------------------------------------------- | ----------- | ------ |
+| 204    | [No Content](https://tools.ietf.org/html/rfc7231#section-6.3.5) | No Content  |        |
+
+To perform this operation, you must be authenticated. [Learn more](authentication.md).
```
I wish I could take credit but it is all automatically generated.  :joy:

code-asher commented on site/src/api/queries/oauth2.ts:
```diff
@@ -0,0 +1,26 @@
+  };
+};
+
+export const oauth2AppSecretsKey = ["oauth-app-secrets"];
```
Ooo I like that, will change.

code-asher commented on site/src/pages/DeploySettingsPage/OAuth2AppsSettingsPage/CreateOAuth2AppPage.tsx:
```diff
@@ -0,0 +1,42 @@
+
+  const postMutation = useMutation({
+    mutationFn: postOAuth2App,
+    onSuccess: (newApp: TypesGen.OAuth2App) => {
```
1. Oh yeah I noticed that too but it felt weird to have the `displaySuccess` stuff in there, however I just realized I can just `await` the mutate and then still do the display in here, I think?  Will fix!
2. Will do!  I was actually not sure if I needed to here because of the navigation; when I navigated back it seemed to have the updated values but is there maybe a case where it might still use the old values?  Even if not necessary, invalidating seems more correct anyway.

code-asher replied:
Huh actually sometimes the icon is missing if I add it and navigate back, probably due to missing invalidation.

code-asher commented on site/src/pages/DeploySettingsPage/OAuth2AppsSettingsPage/CreateOAuth2AppPage.tsx:
```diff
@@ -0,0 +1,42 @@
+      <CreateOAuth2AppPageView
+        isUpdating={postMutation.isLoading}
+        error={postMutation.error}
+        createApp={(req: TypesGen.PostOAuth2AppRequest) => {
```
Ooo yeah you are probably right, I will try removing it.  I probably just added this out of habit.

code-asher commented on site/src/testHelpers/entities.ts:
```diff
@@ -3206,3 +3206,25 @@ export const MockGithubAuthLink: TypesGen.ExternalAuthLink = {
+export const MockOAuth2AppSecrets: TypesGen.OAuth2AppSecret[] = [
+  {
+    id: "1",
+    last_used_at: undefined,
```
Oh I actually did not know that haha.  The backend will just omit the value so I should just remove it altogether, maybe.

code-asher commented on site/src/pages/DeploySettingsPage/OAuth2AppsSettingsPage/CreateOAuth2AppPageView.tsx:
```diff
@@ -0,0 +1,52 @@
+  error,
+}) => {
+  return (
+    <>
```
There are two `Stack`s in here as direct children of the fragment, I think

Parkreiner replied:
Oh, whoops. That's my fault. Misread and thought it was all one stack

code-asher commented on site/src/pages/DeploySettingsPage/OAuth2AppsSettingsPage/CreateOAuth2AppPageView.stories.tsx:
```diff
@@ -0,0 +1,41 @@
+  component: CreateOAuth2AppPageView,
+};
+
+export const Updating = {
```
Good idea!

code-asher commented on site/src/pages/DeploySettingsPage/OAuth2AppsSettingsPage/EditOAuth2AppPage.tsx:
```diff
@@ -0,0 +1,131 @@
+  const queryClient = useQueryClient();
+  const [newAppSecret, setNewAppSecret] = useState<
+    TypesGen.OAuth2AppSecretFull | undefined
+  >(undefined);
```
Oh weird.  I got an error when doing `SetNewAppSecret(undefined)` so I added that.  Could be I was just confused though.

This secret is different from its version in the list of secrets because it has the full secret (you only get to see the full secret once, when the secret is first cretaed).  The list only has truncated secrets, so a cross-reference would not work unfortunately.

code-asher replied:
Oh is it because I did `useState(undefined)` instead of `useState()`?  I will try that.

code-asher replied:
I should call them `fullSecret` and `truncatedSecrets` or something like that, because it is definitely not obvious.

code-asher commented on site/src/pages/DeploySettingsPage/OAuth2AppsSettingsPage/EditOAuth2AppPage.tsx:
```diff
@@ -0,0 +1,131 @@
+    mutationFn: deleteOAuth2App,
+    onSuccess: async () => {
+      displaySuccess(
+        `You have successfully deleted the OAuth2 application "${appName}"`,
```
I omitted the invalidations on all the navigations but I should add them.  I remember there being talk about caching, maybe in those cases the invalidations are required even with navigation?

I decided against a check because I think it is not possible to have an undefined app name here because with no app name there would be no app and the page would still be loading, so no delete button to make a delete request.

That said, I think this is a bad way to do it and maybe this guarantee will not always hold.  According to our conversation above I am going to move these to `queries` and then in the callback I will do:

```
await deleteMutation.mutate(app.id)
displaySuccess(`bla bla ${app.name}`)
```

Which I think will sort it!

code-asher commented on site/src/pages/DeploySettingsPage/OAuth2AppsSettingsPage/EditOAuth2AppPage.tsx:
```diff
@@ -0,0 +1,131 @@
+      deleteOAuth2AppSecret(appId, secretId),
+    onSuccess: async () => {
+      displaySuccess("Successfully deleted an OAuth2 client secret");
+      await queryClient.invalidateQueries([oauth2AppSecretsKey, appId]);
```
Oh wait I though it was supposed to handle all of that for you?  Or do you mean `newAppSecret`?  You have to close the modal to get to a delete button and it is unset when the modal closes, but from a code complete perspective I think clearing it when it matches does make sense.

Parkreiner replied:
Yeah, sorry, I meant the `newAppSecret` value, since it's being stored in separate state

code-asher commented on site/src/pages/DeploySettingsPage/OAuth2AppsSettingsPage/EditOAuth2AppPage.tsx:
```diff
@@ -0,0 +1,131 @@
+                ? "create-secret"
+                : deleteSecretMutation.isLoading
+                  ? "delete-secret"
+                  : false
```
Yeah `isUpdating` is a terrible name, it used to be a boolean and I just never changed the name lol.  I like your suggestion!

So, the multiple resource at once is something that crossed my mind and my original solution was to disable all the other buttons while an update was happening but apparently I got rid of that at some point.

An array sounds like the right solution to me.  Will fix!

code-asher commented on site/src/pages/DeploySettingsPage/OAuth2AppsSettingsPage/EditOAuth2AppPageView.tsx:
```diff
@@ -0,0 +1,308 @@
+  deleteApp: () => void;
+  generateAppSecret: () => void;
+  deleteAppSecret: (id: string) => void;
+  secrets?: TypesGen.OAuth2AppSecret[];
```
Will do!

code-asher commented on site/src/pages/DeploySettingsPage/OAuth2AppsSettingsPage/EditOAuth2AppPageView.tsx:
```diff
@@ -0,0 +1,308 @@
+        <ConfirmDialog
+          hideCancel
+          open={Boolean(newAppSecret)}
+          onConfirm={() => dismissNewSecret()}
```
Oh right, duh!

code-asher commented:
> Github allows any user to make an Oauth2 app. Do we expect this to be the case for Coder?
> Do we want Coder to not allow personal Oauth2 apps, and only allow global oauth apps?
> Should apps be scoped to an org?

I asked and looks like we should be scoping them to an org for the future, and right now owner-only is fine.

code-asher committed:
Add types to oauth2 stories

code-asher committed:
Remove redundant existence checks

Also remove a ? since that condition would short-circuit if `secrets`
was not set.  We could just do a falsey check, but I like to compare
length to a number.

code-asher committed:
Remove unnecessary function wrappers

code-asher committed:
Add types to default meta storybook exports

I think this is right?

code-asher committed:
Mark secrets array as readonly

code-asher committed:
Replace isUpdating with object

This makes it possible to detect parallel mutations.

code-asher committed:
Remove unnecessary undefined on useState

Emyrk commented on coderd/rbac/object.go:
```diff
@@ -198,6 +198,22 @@ var (
 	ResourceTemplateInsights = Object{
 		Type: "template_insights",
 	}
+
+	// ResourceOAuth2App CRUD.
+	//	create/delete = Make or delete an OAuth2 app.
+	//	update = Update the properties of the OAuth2 app.
+	//	read = Read OAuth2 apps.
+	ResourceOAuth2App = Object{
+		Type: "oauth2_app",
+	}
+
+	// ResourceOAuth2AppSecrets CRUD.
+	//	create/delete = Make or delete an OAuth2 app secret.
+	//	update = Update last used date.
+	//	read = Read OAuth2 app hashed or truncated secret.
+	ResourceOAuth2AppSecret = Object{
+		Type: "oauth2_app_secrets",
+	}
```
For now, we can just scope this to the entire platform.

code-asher replied:
Is there something specific I need to do that or is that the default?

Emyrk replied:
No, it is the default. Just recording for posterity.

Emyrk reviewed (commented):
We should name these "Outbound Oauth Apps" or something

Emyrk commented on coderd/httpmw/oauth2.go:
```diff
@@ -178,3 +179,87 @@ func ExtractOAuth2(config OAuth2Config, client *http.Client, authURLOpts map[str
+					Detail:  err.Error(),
+				})
+				return
+			}
```
```suggestion
			if app.ID != secret.AppID {
				httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
					Message: "App ID does not match secret app id.",
				})
				return
			}
```

Emyrk commented on enterprise/coderd/oauth2.go:
```diff
@@ -0,0 +1,238 @@
+	"github.com/coder/coder/v2/cryptorand"
+)
+
+func convertApp(app database.OAuth2App) codersdk.OAuth2App {
+	return codersdk.OAuth2App{
+		ID:          app.ID,
+		Name:        app.Name,
+		CallbackURL: app.CallbackURL,
+		Icon:        app.Icon,
+	}
+}
```
```suggestion
func convertApps(dbApps []database.OAuth2App) []codersdk.OAuth2App {
	apps := []codersdk.OAuth2App{}
	for _, app := range dbApps {
		apps = append(apps, convertApp(app))
	}
	return apps
}

func convertApp(app database.OAuth2App) codersdk.OAuth2App {
	return codersdk.OAuth2App{
		ID:          app.ID,
		Name:        app.Name,
		CallbackURL: app.CallbackURL,
		Icon:        app.Icon,
	}
}
```

Emyrk reviewed (commented):
~~Are we going to want to be OIDC compliant?~~

Eh, this feels like a lot more work, and idk what the upside would be.

Emyrk commented on enterprise/coderd/coderd.go:
```diff
@@ -304,6 +304,31 @@ func New(ctx context.Context, options *Options) (_ *API, err error) {
+						r.Get("/", api.oAuth2AppSecrets)
+						r.Post("/", api.postOAuth2AppSecret)
+
+						r.Route("/{secret}", func(r chi.Router) {
```
```suggestion
						r.Route("/{secret_id}", func(r chi.Router) {
```

Emyrk commented on enterprise/coderd/oauth2.go:
```diff
@@ -0,0 +1,238 @@
+		})
+		return
+	}
+	hashed := sha256.Sum256([]byte(rawSecret))
```
~~Should we salt this with the ID?~~ nvm prob enough randomness in 40 chars

Emyrk commented on enterprise/coderd/oauth2.go:
```diff
@@ -0,0 +1,238 @@
+		ID:            uuid.New(),
+		CreatedAt:     dbtime.Now(),
+		HashedSecret:  hashed[:],
+		DisplaySecret: rawSecret[len(rawSecret)-6:],
```
Make a comment saying this is what github does.

code-asher committed:
Rename apps to "provider apps"

This is to hopefully make it clear these apps are external applications
authorizing with Coder, not the other way around (what we are calling
"external authentication").

Here Coder provides the tokens rather than consuming them.

code-asher committed:
Comment about hashed and display secret

code-asher committed:
Rename secret ID param

This is to make sure it is not mistaken as the actual secret.

code-asher committed:
Move db to sdk conversions to db2sdk.go

code-asher committed:
Be more helpful with app and secret mismatches

code-asher committed:
Clarify new secret vs secret list

code-asher committed:
Remove invalid JSON from test entities

code-asher committed:
Remove redundant type

code-asher committed:
Refactor mutations

Moved them to the queries directory and they all invalidate now.

code-asher committed:
Merge remote-tracking branch 'github/main' into asher/oauth-provider

code-asher committed:
Bump migration number

code-asher committed:
make gen

Forgot to do this after adding a new comment on the apps table.

code-asher committed:
Add dev and entitlement checks to oauth2 provider

The endpoints are disabled until we can implement the token exchange.

Parkreiner commented on site/src/AppRouter.tsx:
```diff
@@ -315,6 +333,15 @@ export const AppRouter: FC = () => {
                 />
+
+                <Route path="oauth2-provider">
+                  <Route path="apps">
```
One thing I noticed during testing: I think there should be a redirect or something when the user tries to go go the `deployment/oauth2-provider/` URL path

Right now, it doesn't take you to a completely empty screen, but you only see the navbar and the Deployment options

code-asher replied:
I am hesitant to do that because one day we might have something else on that page (like for example a toggle for whether to enable the oauth2 provider) and I would not want folks getting confused learning they can go to just `/oauth2-provider` to get to the apps, if that makes sense.

code-asher replied:
That said, seems like it should at least show a proper 404.  Mine gives this error: 
```
"user" must be an existing uuid or username.
queried user="deployment"
```
So I guess it is falling through to some other page.

code-asher replied:
Whoops I was on `dev.coder.com`, yeah you are right it shows a blank page with the deployment sidebar.  That is pretty weird.  Is there even a way for me to show a 404 here?  Maybe a redirect is the only fix...

code-asher replied:
I added a route with `NotFoundPage`.

code-asher replied:
Ah but that is weird too because the sidebar is still there.  At least the page says "not found" I guess.

Emyrk commented:
I'll get a green checkmark on this today. 

We should get this in soon to prevent merge conflicts on the moving base.

Parkreiner commented on site/src/pages/DeploySettingsPage/OAuth2AppsSettingsPage/OAuth2AppsSettingsPageView.tsx:
```diff
@@ -0,0 +1,133 @@
+          description="Configure applications to use Coder as an OAuth2 provider."
+        />
+
+        <Badges>
```
Just to prevent the UI from getting cluttered (especially at smaller screen sizes), I feel like it could help to style the badges like [how we have them on some of the other pages](http://localhost:3000/deployment/appearance)

code-asher replied:
Ohh I see, below the header instead of on the right.  Yeah that looks better, will fix!

Parkreiner commented on site/src/pages/DeploySettingsPage/OAuth2AppsSettingsPage/EditOAuth2AppPageView.tsx:
```diff
@@ -0,0 +1,300 @@
+            <>
+              <p>
+                Your new client secret is displayed below. Make sure to copy it
+                now; you will not be able to see it again.
```
Really like the messaging here; I think I'm going to copy it over to the Tokens tab in the user settings page

Parkreiner commented on site/src/pages/DeploySettingsPage/OAuth2AppsSettingsPage/EditOAuth2AppPageView.tsx:
```diff
@@ -0,0 +1,300 @@
+
+            <Stack direction="row">
+              <div>
+                <strong>Client ID:</strong>
```
I know this was done to make the text bold, but for accessibility/semantics, I'm wondering if it's be better to mark this up as a header, and then style it the same as before

`<strong>` is generally for things that have urgency/seriousness, like warnings

code-asher replied:
Good point!  Will fix.

buenos-nachos reviewed (commented):
The UI worked really well when I was testing things out. Just had a few more comments

code-asher commented on site/src/pages/DeploySettingsPage/OAuth2AppsSettingsPage/EditOAuth2AppPageView.tsx:
```diff
@@ -0,0 +1,300 @@
+              <div>
+                <strong>Client ID:</strong>
+              </div>
+              <CopyableValue value={app.id}>{app.id}</CopyableValue>
```
Oh yeah it is very non-discoverable right now.  I will throw an icon in here.

code-asher committed:
Move badges underneath header

This will work better on smaller screens.

Also I had an extra useless Stack around the button.

code-asher committed:
Improve client ID markup

- Make it a header
- Add a copy icon

code-asher committed:
Show a 404 page on /deployment/oauth2-provider/

One day we might have other settings on this page but for now there is
nothing to show.

code-asher committed:
Clear full new secret if deleted

This should not normally be possible because the modal would prevent
hitting the delete button but you never know.

code-asher committed:
Merge remote-tracking branch 'github/main' into asher/oauth-provider

code-asher committed:
Bump migration number

Emyrk reviewed (approved):
Lets get this in.

code-asher committed:
Fix undefined error in edit oauth2 stories
